airsonic-main/src/main/java/org/airsonic/player/service/SecurityService.java

```diff
@@ -336,15 +374,114 @@ protected boolean isFileInFolder(String file, String folder) {
+    public String getHousehold(String linkCode){
+        Element element = sonosLinkcodeCache.get(linkCode);
+
+        if(element == null){
+            return  null;
+        } else {
+            return (String) element.getValue();
+        }
     }
+
+    /**
+     * Insert the authorisation in sonoslink. Verify is the linkcode exist, and cannot be get again.
+     *
+     * @param username The username authorisation for sonos link
+     * @param householdId The id of entry in sonos controller
+     * @param linkcode The link code used between sonos and airsonic.
+     * @return true if the insert is ok, false if some entry exist with the linkcode
+     */
+    public boolean authoriseSonos(String username, String householdId, String linkcode) {
+        if(sonosLinkDao.findByLinkcode(linkcode) != null){
+            return false;
+        }
+
+        sonosLinkDao.create(new SonosLink(username, householdId, linkcode));
+        return true;
+    }
+
+    /**
+     * Find the user they have a linkCode for the householdId set, build the authToken and return it.
+     *
+     * @param householdId The householdId from Sonos
+     * @param linkCode The linkCode return it before
+     * @return The build authToken or null if didn't find any user with householdId and linkCode
+     */
+    public String getSonosAuthToken(String householdId, String linkCode){
+
+        SonosLink sonosLink = sonosLinkDao.findByLinkcode(linkCode);
+        if(sonosLink != null && householdId.equals(sonosLink.getHouseholdid())) {
+            return buildToken(sonosLink);
+        } else {
+            return null;
+        }
+    }
+
+    private String buildToken(SonosLink link) {
+        return jwtSecurityService.createSonosToken(link.getUsername(), link.getHouseholdid(), link.getLinkcode());
+    }
+
+    public SonosLink getSonosLink(String linkCode) {
+        return sonosLinkDao.findByLinkcode(linkCode);
+    }
+
+    public void authenticate(String sonosLinkToken) throws SonosSoapFault.LoginUnauthorized {
+        SonosLink sonosLink =jwtSecurityService.verifySonosLink(sonosLinkToken);
+
+        SonosLink saved = sonosLinkDao.findByLinkcode(sonosLink.getLinkcode());
+        if(saved != null && saved.identical(sonosLink)){
+            setUser(sonosLink.getUsername());
+        } else {
+            throw new SonosSoapFault.LoginUnauthorized();
+        }
+    }
+
+    public void authenticate() throws SonosSoapFault.LoginUnauthorized {
+        setUser(USERNAME_SONOS);
+    }
+
+    private void setUser(String username) throws SonosSoapFault.LoginUnauthorized {
+        User user = getUserByName(USERNAME_SONOS, true);
+        Authentication authentication = authenticate(user.getUsername(), user.getPassword());
+        SecurityContextHolder.getContext().setAuthentication(authentication);
+    }
+
+        // must be strict 32 character limit on length
+    private String createLinkCode() {
+        return UUID.randomUUID().toString().replace("-", "");
```

Comment:
If this is security-related, please use [SecureRandom]( https://docs.oracle.com/javase/8/docs/api/java/security/SecureRandom.html ) instead.

Reply:
Right.